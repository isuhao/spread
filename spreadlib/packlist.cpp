#include "packlist.hpp"
#include <map>
#include <stdexcept>
#include <misc/readjson.hpp>
#include "infojson.hpp"

using namespace Spread;

typedef std::map<std::string,PackInfo*> Lookup;

struct PackList::_Internal
{
  PackInfoList list;
  Lookup lookup;
};

PackList::PackList() { ptr.reset(new _Internal); }

PackInfoList PackList::getList() const { return ptr->list; }

static void fail(const std::string &msg)
{ throw std::runtime_error(msg); }

void PackList::loadJson(const std::string &file, const std::string &channel)
{
  assert(file != "");
  clear();

  using namespace Json;

  Value root = ReadJson::readJson(file);

  if(!root.isObject()) fail("Invalid pack file " + file);

  Value::Members keys = root.getMemberNames();
  ptr->list.resize(keys.size());
  for(int i=0; i<keys.size(); i++)
    {
      const std::string &key = keys[i];
      PackInfo &p = ptr->list[i];
      ptr->lookup[key] = &p;
      jsonToInfo(p, root[key], channel, key);
    }
}

PackInfo PackList::get(const std::string &pack) const
{
  Lookup::const_iterator it = ptr->lookup.find(pack);
  if(it == ptr->lookup.end())
    fail("Unknown package " + pack);
  return *it->second;
}

void PackList::clear()
{
  ptr->list.clear();
  ptr->lookup.clear();
}
