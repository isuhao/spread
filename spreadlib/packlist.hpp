#ifndef __SPREADLIB_PACKLIST_HPP_
#define __SPREADLIB_PACKLIST_HPP_

#include "packinfo.hpp"
#include <boost/shared_ptr.hpp>

/* This is the class responsible for loading and holding the
   packs.json data.
 */

namespace Spread
{
  struct PackList
  {
    PackList();

    // Return list of all packages
    const PackInfoList &getList() const;

    // Get one package, throw if it doesn't exist
    const PackInfo &get(const std::string &pack) const;

    /* Load packlist from a json file. The PackList is reusable, so
       loadJson() may be called multiple times. Each call clears out
       all existing data before loading the new information.
     */
    void loadJson(const std::string &file, const std::string &channel);

    /* Clear out all existing pack data, reverting the object back to
       the same state as right after it was constructed.
     */
    void clear();

  private:
    struct _Internal;
    boost::shared_ptr<_Internal> ptr;
  };
};

#endif
