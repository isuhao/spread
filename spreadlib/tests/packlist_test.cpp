#include "packlist.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

void print(const PackInfo &pack)
{
  cout << "PACK: " << pack.package
       << "\nChannel: " << pack.channel
       << "\nVersion: " << pack.version
       << "\nSize: " << pack.installSize
       << "\nDirs:\n";
  assert(pack.dirs.size() == pack.paths.size());
  for(int i=0; i<pack.dirs.size(); i++)
    cout << "  " << pack.dirs[i] << "   " << pack.paths[i] << endl;
  cout << endl;
}

void print(const PackInfoList &list)
{
  cout << "ALL PACKS:\n";
  for(int i=0; i<list.size(); i++)
    print(list[i]);
}

int main()
{
  PackList list("test");
  list.loadJson("data_packs1.json");
  print(list.get("shots300x260"));
  try { list.get("error"); }
  catch(exception &e) { cout << "Expected exception: " << e.what() << endl; }
  print(list.getList());
  list.loadJson("data_packs2.json");
  print(list.getList());
  list.clear();
  print(list.getList());

  return 0;
}
