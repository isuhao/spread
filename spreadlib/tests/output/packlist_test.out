PACK: shots300x260
Channel: test
Version: 
Size: 0
Dirs:
  BR60SIN210HuA5oZHmAJGSH37cwdw0vDf-DqB8acv82TUw   
  RUwhlaWTmUQvb8mLi_4d-pIZIq1aXfP8H5GrdcR5wJTdCg   
  Mn57JWXqrcRWVjpslQs8Odx7GzXiwfQGwlNUUC__7VLYCg   

Expected exception: Unknown package error
ALL PACKS:
PACK: empty
Channel: test
Version: 
Size: 0
Dirs:

PACK: shots300x260
Channel: test
Version: 
Size: 0
Dirs:
  BR60SIN210HuA5oZHmAJGSH37cwdw0vDf-DqB8acv82TUw   
  RUwhlaWTmUQvb8mLi_4d-pIZIq1aXfP8H5GrdcR5wJTdCg   
  Mn57JWXqrcRWVjpslQs8Odx7GzXiwfQGwlNUUC__7VLYCg   

PACK: tiggit.net/10-min-space-strategy
Channel: test
Version: 1.03
Size: 0
Dirs:
  iLmpxkzLr_OSwwnBw0ABoAO95RruG3x1kcvs3bNgHV5RC   

PACK: tiggit.net/1213
Channel: test
Version: 
Size: 4321
Dirs:
  feFIKaST1fhBKHg_OkzqQDuGgTpkoKavsOC2xfNu2H0TAg   some/path

PACK: tiggit.net/blocktality
Channel: test
Version: 1.0.1
Size: 1234567
Dirs:
  qIcK0UT8YRao378Cock7NauduHEDHHw6Ic23pdkqpGckQg   

PACK: wxtiggit1-win32
Channel: test
Version: 0.88 beta
Size: 0
Dirs:
  SYO4ROrGazq43ogPyUW33XgjsMJ-3AaYfCQIQenHD748   
  bwhLyp9hTYpUs_yC7nJcoiv3f2zoRBvcA3dwlxVsFyc8AQ   

ALL PACKS:
PACK: shots300x260
Channel: test
Version: 
Size: 0
Dirs:
  BR60SIN210HuA5oZHmAJGSH37cwdw0vDf-DqB8acv82TUw   
  RUwhlaWTmUQvb8mLi_4d-pIZIq1aXfP8H5GrdcR5wJTdCg   
  Mn57JWXqrcRWVjpslQs8Odx7GzXiwfQGwlNUUC__7VLYCg   
  Mn57JWXqrcRWVjpslQs8Odx7GzXiwfQGwlNUUC__7VLYCg   

PACK: tiggit.net/10-min-space-strategy
Channel: test
Version: 1.10
Size: 0
Dirs:
  iLmpxdfadfdfdfdafadfAO95RruG3x1kcvs3bNgHV5RC   

PACK: tiggit.net/newgame
Channel: test
Version: 1.0.1
Size: 1234567
Dirs:
  qIcK0UT8YRao378Cock7NauduHEDHHw6Ic23pdkqpGckQg   

PACK: wxtiggit1-win32
Channel: test
Version: 0.88 beta
Size: 0
Dirs:
  SYO4ROrGazq43ogPyUW33XgjsMJ-3AaYfCQIQenHD748   
  bwhLyp9hTYpUs_yC7nJcoiv3f2zoRBvcA3dwlxVsFyc8AQ   

ALL PACKS:
